## Groceries Market Basket Dataset

The dataset contains supermarket transactions, where each row represents a single transaction and each column represents a specific item. A value of True indicates that the item was purchased in that transaction, while False indicates it was not purchased.

First, we import the libraries:

*   Pandas - for data manipulation.
*   MatPlotLib - for data visualization.
*   Seaborn - for data visualization.
*   MLXtend - to apply the Apriori Algorithm.

In [2]:
# importing the libraries
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from matplotlib import pyplot as plt
import seaborn as sns

We load the dataset into a Pandas DataFrame and view the first few rows to inspect the structure.

In [6]:
# loading the dataset
df = pd.read_csv('input/sol1.csv', index_col=0)
df.head()

Apple  Bread  Butter  Cheese   Corn   Dill   Eggs  Ice cream  Kidney Beans  \
0  False   True   False   False   True   True  False       True         False   
1  False  False   False   False  False  False  False      False         False   
2   True  False    True   False  False   True  False       True         False   
3  False  False    True    True  False   True  False      False         False   
4   True   True   False   False  False  False  False      False         False   

    Milk  Nutmeg  Onion  Sugar  Unicorn  Yogurt  chocolate  
0  False   False  False   True    False    True       True  
1   True   False  False  False    False   False      False  
2   True   False  False  False    False    True       True  
3   True    True   True  False    False   False      False  
4  False   False  False  False    False   False      False

#### **Q1** How many transactions and items are there in the data set?


We check the number of transactions and the number of items in the dataset. Rows correspond to transactions, and columns correspond to items.

In [7]:
# finding the dimensions of the dataframe
df.shape

(999, 16)

As we can see, there are 999 rows, meaning 999 transactions, and 16 columns, meaning 16 items in the dataset.

To prepare the data for the following questions, we apply the Apriori algorithm on the dataframe and set the minimum support parameter to 2%.

In [8]:
# applying the apriori algorithm
frequent_itemsets = apriori(df, min_support=0.02, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets

support                                           itemsets  length
0     0.383383                                            (Apple)       1
1     0.384384                                            (Bread)       1
2     0.420420                                           (Butter)       1
3     0.404404                                           (Cheese)       1
4     0.407407                                             (Corn)       1
...        ...                                                ...     ...
5540  0.020020  (Ice cream, Nutmeg, Cheese, Kidney Beans, Onio...       6
5541  0.021021  (Unicorn, Cheese, chocolate, Kidney Beans, Oni...       6
5542  0.020020  (Sugar, Ice cream, Nutmeg, Cheese, chocolate, ...       6
5543  0.020020  (Yogurt, Nutmeg, Corn, Unicorn, chocolate, Onion)       6
5544  0.020020  (Ice cream, Milk, Unicorn, chocolate, Onion, D...       6

[5545 rows x 3 columns]

#### **Q4** Find top selling items with minimum support of 5%.

To solve Question 4, first we sort the dataframe by support in the descending order by using the sort_values() function from the Pandas library and setting the by and ascending parameters to support and False respectively.

In [9]:
# sorting the dataframe
frequent_itemsets = frequent_itemsets.sort_values(by='support', ascending=False)

Next, we filter the dataframe to find items with length 1 and support more than equal to 5%. Then we splice the sorted and filtered dataframe to show only the top 5 entries.

In [12]:
# finding top 5 items with minimum support of 5%
frequent_itemsets[ (frequent_itemsets['length'] == 1) &
                   (frequent_itemsets['support'] >= 0.05) ][0:5]

support     itemsets  length
15  0.421421  (chocolate)       1
2   0.420420     (Butter)       1
14  0.420420     (Yogurt)       1
7   0.410410  (Ice cream)       1
12  0.409409      (Sugar)       1

As we can see, chocolate, butter, yogurt, ice cream and sugar are the top 5 selling items with support of 42%, 42%, 42%, 41%, and 40% respectively.

#### **Q5.** Find all frequent itemsets with minimum support of 20%.

To solve Question 5, we filter the dataframe to find itemsets having length more than 1, and support more than 5%.

In [ ]:
# finding itemsets having length more than 1 and minimum support of 20%
frequent_itemsets[(frequent_itemsets['length'] > 1) & 
                  (frequent_itemsets['support'] >= 0.20)]

support                itemsets  length
120  0.211211       (Milk, chocolate)       2
49   0.207207     (Ice cream, Butter)       2
107  0.202202  (chocolate, Ice cream)       2
50   0.202202  (Butter, Kidney Beans)       2
57   0.202202     (chocolate, Butter)       2
62   0.200200  (Kidney Beans, Cheese)       2

As we can see, there are only 6 itemsets, having support of around 20%.

#### **Q6.**  Find all frequent itemsets of length 2 with minimum support of 19%.

To solve Question 6, we filter the dataframe to find itemsets having length 2 and minimum support of 19%.

In [ ]:
# finding itemsets having length 2 and minimum support of 19%
frequent_itemsets[(frequent_itemsets['length'] == 2) & 
                  (frequent_itemsets['support'] >= 0.19)]

support                   itemsets  length
120  0.211211          (Milk, chocolate)       2
49   0.207207        (Ice cream, Butter)       2
107  0.202202     (chocolate, Ice cream)       2
50   0.202202     (Butter, Kidney Beans)       2
57   0.202202        (chocolate, Butter)       2
62   0.200200     (Kidney Beans, Cheese)       2
90   0.199199          (chocolate, Dill)       2
108  0.199199       (Milk, Kidney Beans)       2
52   0.198198           (Butter, Nutmeg)       2
51   0.198198             (Milk, Butter)       2
135  0.198198        (chocolate, Yogurt)       2
53   0.197197            (Butter, Onion)       2
54   0.196196            (Sugar, Butter)       2
100  0.196196  (Ice cream, Kidney Beans)       2
129  0.196196         (chocolate, Onion)       2
121  0.195195            (Onion, Nutmeg)       2
73   0.195195       (Corn, Kidney Beans)       2
104  0.195195         (Sugar, Ice cream)       2
113  0.194194     (Kidney Beans, Yogurt)       2
43   0.193193            (Yogurt, Bread)       2
74   0.193193               (Milk, Corn)       2
122  0.193193            (Sugar, Nutmeg)       2
80   0.192192          (chocolate, Corn)       2
103  0.192192         (Ice cream, Onion)       2
64   0.192192           (Cheese, Nutmeg)       2
86   0.192192              (Dill, Onion)       2
72   0.192192          (Corn, Ice cream)       2
124  0.192192           (Yogurt, Nutmeg)       2
128  0.192192            (Yogurt, Onion)       2
131  0.191191            (Sugar, Yogurt)       2
56   0.191191           (Butter, Yogurt)       2
114  0.191191  (chocolate, Kidney Beans)       2
46   0.191191             (Corn, Butter)       2
119  0.190190             (Milk, Yogurt)       2
84   0.190190               (Milk, Dill)       2
79   0.190190             (Corn, Yogurt)       2
126  0.190190             (Sugar, Onion)       2

As we can see, there are 37 itemsets having length 2 with support more than or equal to 19%. The support ranges between 21% and 19% with milk and chocolate having the highest support, and sugar and onion having the minimum support.

#### **Q7.** Find the top 10 association rules with minimum support of 20%, sorted by confidence in descending order.


To solve Question 7, we first find the association rules using the association_rules() function from the MLXtend library and set the parameter metric to support, and the min_threshold to 20%.

In [29]:
# finding top 10 association rules with minimum support of 19%
rules = association_rules(frequent_itemsets, metric='support', min_threshold=0.20)
rules

antecedents     consequents  antecedent support  consequent support  \
0           (Milk)     (chocolate)            0.405405            0.421421   
1      (chocolate)          (Milk)            0.421421            0.405405   
2      (Ice cream)        (Butter)            0.410410            0.420420   
3         (Butter)     (Ice cream)            0.420420            0.410410   
4      (chocolate)     (Ice cream)            0.421421            0.410410   
5      (Ice cream)     (chocolate)            0.410410            0.421421   
6         (Butter)  (Kidney Beans)            0.420420            0.408408   
7   (Kidney Beans)        (Butter)            0.408408            0.420420   
8      (chocolate)        (Butter)            0.421421            0.420420   
9         (Butter)     (chocolate)            0.420420            0.421421   
10  (Kidney Beans)        (Cheese)            0.408408            0.404404   
11        (Cheese)  (Kidney Beans)            0.404404            0.408408   

     support  confidence      lift  representativity  leverage  conviction  \
0   0.211211    0.520988  1.236263               1.0  0.040365    1.207857   
1   0.211211    0.501188  1.236263               1.0  0.040365    1.192021   
2   0.207207    0.504878  1.200889               1.0  0.034662    1.170579   
3   0.207207    0.492857  1.200889               1.0  0.034662    1.162571   
4   0.202202    0.479810  1.169098               1.0  0.029246    1.133412   
5   0.202202    0.492683  1.169098               1.0  0.029246    1.140467   
6   0.202202    0.480952  1.177626               1.0  0.030499    1.139764   
7   0.202202    0.495098  1.177626               1.0  0.030499    1.147905   
8   0.202202    0.479810  1.141262               1.0  0.025028    1.114169   
9   0.202202    0.480952  1.141262               1.0  0.025028    1.114693   
10  0.200200    0.490196  1.212143               1.0  0.035038    1.168284   
11  0.200200    0.495050  1.212143               1.0  0.035038    1.171583   

    zhangs_metric   jaccard  certainty  kulczynski  
0        0.321413  0.343089   0.172088    0.511088  
1        0.330310  0.343089   0.161088    0.511088  
2        0.283728  0.332263   0.145722    0.498868  
3        0.288629  0.332263   0.139837    0.498868  
4        0.249991  0.321145   0.117708    0.486246  
5        0.245323  0.321145   0.123167    0.486246  
6        0.260247  0.322684   0.122625    0.488025  
7        0.254963  0.322684   0.128848    0.488025  
8        0.213933  0.316119   0.102470    0.480381  
9        0.213564  0.316119   0.102892    0.480381  
10       0.295838  0.326797   0.144043    0.492623  
11       0.293849  0.326797   0.146454    0.492623

Then we sort the generated association rules in the descending order by confidence by using the sort_values() function from the Pandas library and setting the by and ascending parameters to confidence and False respectively. Then we splice the sorted dataframe to show the top 10 rules.

In [30]:
# sorting the rules in the descending order by confidence
rules.sort_values(by='confidence', ascending=False)[0:10]

antecedents     consequents  antecedent support  consequent support  \
0           (Milk)     (chocolate)            0.405405            0.421421   
2      (Ice cream)        (Butter)            0.410410            0.420420   
1      (chocolate)          (Milk)            0.421421            0.405405   
7   (Kidney Beans)        (Butter)            0.408408            0.420420   
11        (Cheese)  (Kidney Beans)            0.404404            0.408408   
3         (Butter)     (Ice cream)            0.420420            0.410410   
5      (Ice cream)     (chocolate)            0.410410            0.421421   
10  (Kidney Beans)        (Cheese)            0.408408            0.404404   
6         (Butter)  (Kidney Beans)            0.420420            0.408408   
9         (Butter)     (chocolate)            0.420420            0.421421   

     support  confidence      lift  representativity  leverage  conviction  \
0   0.211211    0.520988  1.236263               1.0  0.040365    1.207857   
2   0.207207    0.504878  1.200889               1.0  0.034662    1.170579   
1   0.211211    0.501188  1.236263               1.0  0.040365    1.192021   
7   0.202202    0.495098  1.177626               1.0  0.030499    1.147905   
11  0.200200    0.495050  1.212143               1.0  0.035038    1.171583   
3   0.207207    0.492857  1.200889               1.0  0.034662    1.162571   
5   0.202202    0.492683  1.169098               1.0  0.029246    1.140467   
10  0.200200    0.490196  1.212143               1.0  0.035038    1.168284   
6   0.202202    0.480952  1.177626               1.0  0.030499    1.139764   
9   0.202202    0.480952  1.141262               1.0  0.025028    1.114693   

    zhangs_metric   jaccard  certainty  kulczynski  
0        0.321413  0.343089   0.172088    0.511088  
2        0.283728  0.332263   0.145722    0.498868  
1        0.330310  0.343089   0.161088    0.511088  
7        0.254963  0.322684   0.128848    0.488025  
11       0.293849  0.326797   0.146454    0.492623  
3        0.288629  0.332263   0.139837    0.498868  
5        0.245323  0.321145   0.123167    0.486246  
10       0.295838  0.326797   0.144043    0.492623  
6        0.260247  0.322684   0.122625    0.488025  
9        0.213564  0.316119   0.102892    0.480381

#### **Q8.** Find association rules with minimum support of 20% and lift of more than 1.0.


To solve Question 8, we filter the dataframe to have lift more than 1.

In [ ]:
# finding association rules with minimum support of 20% and having lift more than 1
rules[(rules['support'] >= 0.20) &
      (rules['lift'] > 1.0)]

antecedents     consequents  antecedent support  consequent support  \
0           (Milk)     (chocolate)            0.405405            0.421421   
1      (chocolate)          (Milk)            0.421421            0.405405   
2      (Ice cream)        (Butter)            0.410410            0.420420   
3         (Butter)     (Ice cream)            0.420420            0.410410   
4      (chocolate)     (Ice cream)            0.421421            0.410410   
5      (Ice cream)     (chocolate)            0.410410            0.421421   
6         (Butter)  (Kidney Beans)            0.420420            0.408408   
7   (Kidney Beans)        (Butter)            0.408408            0.420420   
8      (chocolate)        (Butter)            0.421421            0.420420   
9         (Butter)     (chocolate)            0.420420            0.421421   
10  (Kidney Beans)        (Cheese)            0.408408            0.404404   
11        (Cheese)  (Kidney Beans)            0.404404            0.408408   

     support  confidence      lift  representativity  leverage  conviction  \
0   0.211211    0.520988  1.236263               1.0  0.040365    1.207857   
1   0.211211    0.501188  1.236263               1.0  0.040365    1.192021   
2   0.207207    0.504878  1.200889               1.0  0.034662    1.170579   
3   0.207207    0.492857  1.200889               1.0  0.034662    1.162571   
4   0.202202    0.479810  1.169098               1.0  0.029246    1.133412   
5   0.202202    0.492683  1.169098               1.0  0.029246    1.140467   
6   0.202202    0.480952  1.177626               1.0  0.030499    1.139764   
7   0.202202    0.495098  1.177626               1.0  0.030499    1.147905   
8   0.202202    0.479810  1.141262               1.0  0.025028    1.114169   
9   0.202202    0.480952  1.141262               1.0  0.025028    1.114693   
10  0.200200    0.490196  1.212143               1.0  0.035038    1.168284   
11  0.200200    0.495050  1.212143               1.0  0.035038    1.171583   

    zhangs_metric   jaccard  certainty  kulczynski  
0        0.321413  0.343089   0.172088    0.511088  
1        0.330310  0.343089   0.161088    0.511088  
2        0.283728  0.332263   0.145722    0.498868  
3        0.288629  0.332263   0.139837    0.498868  
4        0.249991  0.321145   0.117708    0.486246  
5        0.245323  0.321145   0.123167    0.486246  
6        0.260247  0.322684   0.122625    0.488025  
7        0.254963  0.322684   0.128848    0.488025  
8        0.213933  0.316119   0.102470    0.480381  
9        0.213564  0.316119   0.102892    0.480381  
10       0.295838  0.326797   0.144043    0.492623  
11       0.293849  0.326797   0.146454    0.492623